zokrates_core/src/flatten/mod.rs

```diff
@@ -620,59 +664,27 @@ impl<'ast, T: Field> Flattener<'ast, T> {
         e: FlatExpression<T>,
         c: T,
     ) -> FlatExpression<T> {
-        let bit_width = T::get_required_bits();
-        // decompose e to bits
-        let e_id = self.define(e, statements_flattened);
-
-        // define variables for the bits
-        let e_bits_be: Vec<FlatVariable> = (0..bit_width).map(|_| self.use_sym()).collect();
-
-        // add a directive to get the bits
-        statements_flattened.push_back(FlatStatement::Directive(FlatDirective::new(
-            e_bits_be.clone(),
-            Solver::bits(bit_width),
-            vec![e_id],
-        )));
-
-        // bitness checks
-        for bit in e_bits_be.iter().take(bit_width) {
-            statements_flattened.push_back(FlatStatement::Condition(
-                FlatExpression::Identifier(*bit),
-                FlatExpression::Mult(
-                    box FlatExpression::Identifier(*bit),
-                    box FlatExpression::Identifier(*bit),
-                ),
-                RuntimeError::ConstantLtBitness,
-            ));
-        }
-
-        // bit decomposition check
-        let mut e_sum = FlatExpression::Number(T::from(0));
+        let e: FlatExpression<T> = self.define(e, statements_flattened).into();
+        let bitwidth = T::get_required_bits();
 
-        for (i, bit) in e_bits_be.iter().take(bit_width).enumerate() {
-            e_sum = FlatExpression::Add(
-                box e_sum,
-                box FlatExpression::Mult(
-                    box FlatExpression::Identifier(*bit),
-                    box FlatExpression::Number(T::from(2).pow(bit_width - i - 1)),
-                ),
-            );
-        }
-
-        statements_flattened.push_back(FlatStatement::Condition(
-            FlatExpression::Identifier(e_id),
-            e_sum,
+        let e_bits_be = self.get_bits(
+            &FlatUExpression::with_field(e),
+            bitwidth,
+            bitwidth,
+            statements_flattened,
             RuntimeError::ConstantLtSum,
-        ));
+        );
 
         // check that this decomposition does not overflow the field
-        self.enforce_constant_le_check(
+        self.enforce_constant_le_check_bits(
```

Comment:
Can we use `enforce_constant_le_check` directly ?

Reply:
Hmm, we need bits of `e` after this call so I think it's fine as is, if we call it directly then we are doing unnecessarily calls and potentially cloning stuff.

Reply:
I moved this to `get_bits`: if you call `get_bits` and want a decomposition of max bitwidth, you always need that check because of non determinism.

Reply:
Ok now I understand it's not possible, as in some cases it would mean checking both `< P` (in `get_bits`) and `< C` (for a `< C` check)

Reply:
yeah